druid/examples/game_of_life.rs

```diff
@@ -0,0 +1,171 @@
+
+
+const POOL_SIZE: i32 = 1600;
+const CELL_COLOR: Color = Color::rgb8(0xf3 as u8, 0xf4 as u8, 8 as u8);
+
+#[derive(Clone)]
+struct AppData {
+    grid: [i32; 1600],
+}
+
+impl AppData {
+    pub fn evolve(&mut self) {
+        let mut indices_to_mutate: Vec<usize> = vec![];
+        for i in 0..POOL_SIZE {
+            let neighbors = AppData::neighbors(i);
+            let mut sum = 0;
+            for nb_i in &neighbors {
+                if *nb_i >= 0 && *nb_i < POOL_SIZE {
+                    let idx = *nb_i as usize;
+                    let neighbor_life = self.grid[idx];
+                    if neighbor_life > 0 {
+                        sum += neighbor_life;
+                    }
+                }
+            }
+            let life = self.grid[i as usize];
+            // death by loneliness or overcrowding
+            if life == 1 && (sum < 2 || sum > 3) {
+                indices_to_mutate.push(i as usize);
+                continue;
+            }
+            if life == -1 && sum == 3 {
+                indices_to_mutate.push(i as usize);
+            }
+        }
+        println!("mut idx: {:?}", indices_to_mutate);
+        for i_mut in indices_to_mutate {
+            self.grid[i_mut] *= -1;
+        }
+    }
+
+    // return the indices of neighbors of a cell, clockwise, starting from top-left corner
+    pub fn neighbors(i: i32) -> [i32; 8]{
+        return [i-41, i-40, i-39, i+1, i+41, i+40, i+39, i-1];
+    }
+}
+
+impl Data for AppData {
+    fn same(&self, other: &Self) -> bool {
+        for i in 0..POOL_SIZE {
+            if self.grid[i as usize] != other.grid[i as usize] {
+                return false;
+            }
+        }
+        return true;
+    }
+}
+
+struct CustomWidget;
+
+
+impl Widget<AppData> for CustomWidget {
+    fn event(&mut self, _ctx: &mut EventCtx, _event: &Event, _data: &mut AppData, _env: &Env) {
+        match _event {
+            Event::MouseDown(_e) => {
+                _data.evolve();
+                _ctx.request_paint();
+            }
+            _ => {}
+        }
+    }
+
+    fn lifecycle(
+        &mut self,
+        _ctx: &mut LifeCycleCtx,
+        _event: &LifeCycle,
+        _data: &AppData,
+        _env: &Env,
+    ) {}
+
+    fn update(&mut self, _ctx: &mut UpdateCtx, _old_data: &AppData, _data: &AppData, _env: &Env) {}
+
+    fn layout(
+        &mut self,
+        _layout_ctx: &mut LayoutCtx,
+        bc: &BoxConstraints,
+        _data: &AppData,
+        _env: &Env,
+    ) -> Size {
+        bc.max()
+    }
+
+    fn paint(&mut self, paint_ctx: &mut PaintCtx, data: &AppData, _env: &Env) {
+        let size: Size = paint_ctx.size();
+        let w0 = size.width/40.0;
+        let h0 = size.height/40.0;
+        let cell_size = Size{ width: w0, height: h0 };
+        for i in 0..1600 {
```

Comment:
and a general rust thought: I would probably (in a "real" program) write this as using an iterator returned from my `Grid` type; there would be a function like,

```rust
iter_live_cells(&self) -> impl Iterator<Item=(usize, usize)> { .. }
```
and then here it would just be like,

```rust
for (row, col) in grid.iter_live_cells() {
// make rect
// paint
}
```